### DISCUSSION TITLE
window placement

### DISCUSSION THREAD

**Discussion by JaapStruyk** [Schneegans/Fly-Pie#161] [Ideas & Feature Discussions]
Is it possible to add window placement request to the started application like with xdotool so it will be placed on the screen where it was started with fly-pie?
In my case I use mouse gesture to start fly-pie (which is placed on the screen where I make the gesture) and it would be very intuitive if the started application would be placed on that same location.

**Comment by Schneegans** [comment]
Thanks for the suggestion! However, I am not really sure about this. On the one hand side, I understand that it would be more intuitive (and faster as it will require less mouse travel to interact with the newly opened app). However, on the other hand there are some difficulties:
* I did some quick tests and failed to get it working at all. It still might be possible, but moving windows around is something which is in the scope of Mutter and I am not sure if it's possible for GNOME Shell extensions (especially under Wayland). Applications can give some hints to the compositor where windows should be opened, but this can be ignored by the compositor. I'll have to test some more.
* If applications are opened with the "Launch Application" action, we have no reliable way to know which windows got opened by the command. For example, your action could execute a shell script which opens twenty applications or maybe runs only some command in the background. The only thing I could think of is to position _any_ window which is opened, let's say, within one second after the execution of the action at the mouse pointer.

What do you think?

**Comment by JaapStruyk** [comment]
Thanks for looking at it!
The reason for the idea was my brand new 49" monitor and the miles off mouse travel I didn't anticipated on ;-)

Using xdotool window placement after starting an app works very well, don't know about Wayland but under X-gnome40 working very well with `xdotool search --onlyvisible --name firefox windowmove 10 10` even when using numbers way out off the screen like 10000 the window is placed at the edge of the visible screen but didn't find a way to pass placement wishes to the compositor on startup off any application. (which is a mess on gnome anyway, some apps remember placement others behave in an anarchistic way)
xdotool command could be called after a fixed delay (like U suggest)  extracting the bins name from the called desktop file but that won't work for scripts containing others script etc. apart from if this behaviour is wanted as standard action anyway.
The one second (or adjustable in case off slow opening apps) isn't a show stopper I think, it's always faster then dragging the window to the desired place.

Thanks again for looking, think this would make a very intuitive user desktop experience.

**Comment by Schneegans** [comment]
I played around a bit more and discovered a pretty reliable solution! Fly-Pie now checks whether a new window is opened within one second after an action has been executed. If it detects a new window, it checks whether the new window got focused (i.e. it did not open in the background). If this is the case it centers the window below the pointer (clamped to the current monitor bounds so that nothing goes off-screen). If this does not happen within one second, nothing is done.

You can easily test the new behavior by [installing the current development version of Fly-Pie](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/installation.md#c-cloning-the-latest-version-with-git)! (Feel free to ask is anything is unclear in this guide).

I would really appreciate some feedback as there might be some weird behavior with apps messing around with window positioning on their own...

**Comment by JaapStruyk** [comment]
Just tried and works very very well!!! Really cool!
Not all apps respond to the placement request, java based programs (Arduino Fritzing), appimage programs and some others started with a shell script pointing to the binary's location (Qcad TunerstudioMS) don't respond. Not sure id this is window manager, will make some more tests this weekend.
But in general great work! (and a nice way off coding it!)

**Comment by Schneegans** [comment]
I encountered some applications which take longer than one second to open a window (Java based programs may also take a bit longer, I guess). If you want to, you could increase the timeout ([line 714 in `common/utils.js`](https://github.com/Schneegans/Fly-Pie/blob/eb6d1552d3055fe8af8c83c51c574c5ebcd69a52/src/common/utils.js#L714)) and test if that helps. If we increase this, we should maybe store the pointer location when the action was activated and use this position later to position the window once it is opened. This is because the user may have moved the pointer somewhere else in the meantime. I guess, it would feel weird if the window opened at this new pointer location.

**Comment by JaapStruyk** [comment]
Bingo! With timeout at 2000ms all apps open at mouse pointers location and indeed storing the mouse pointers location is a good idea, when I move it within the 2 seconds window follows pointer location. (one could see that as feature instead as bug ;-)
Overall working awesome! This is so darn cool!!!

**Comment by Schneegans** [comment]
Great that it works for you! I just increased the timeout to 2000ms as you suggested and stored the pointer location. Just do a `git pull origin develop` followed by a `make install` to test it (maybe you have to undo your changes before).

Thanks again for this suggestion, I already love this behavior!

**Comment by JaapStruyk** [comment]
Working awesome! Thanks for implementing!!!
